# import data

In [41]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [42]:
file_path = 'C:/Users/Demi/Desktop/TB2/Mini project/df.csv'
data = pd.read_csv(file_path)
data = data.drop('Unnamed: 0', axis=1)
data

Account No       Balance   Amount  Third Party Account No  \
0       678330503.0   2971.000000  1584.00                     NaN   
1       472213568.0   3792.000000  1950.00                     NaN   
2       472213568.0   3012.000000  -780.00             283027736.0   
3       283027736.0   1787.000000   780.00             472213568.0   
4       624500124.0   3226.000000  1825.00                     NaN   
...             ...           ...      ...                     ...   
230190  581655972.0  45935.206861   -41.06                     NaN   
230191  786141370.0   -244.837500   -62.35                     NaN   
230192  824916823.0   9709.172159   -32.94                     NaN   
230193  366550080.0  26834.165794   -19.25                     NaN   
230194  259015404.0   1145.349837   -37.64                     NaN   

          Third Party Name             DateTime  
0       Westport Care Home  2023-01-01 00:00:00  
1         Barbiee Boutique  2023-01-01 00:00:00  
2                      NaN  2023-01-01 00:00:00  
3                      NaN  2023-01-01 00:00:00  
4                 Fat Face  2023-01-01 00:00:00  
...                    ...                  ...  
230190               Tesco  2023-12-06 20:54:00  
230191     Sainsbury Local  2023-12-06 20:55:00  
230192           Deliveroo  2023-12-06 21:05:00  
230193              Amazon  2023-12-06 21:13:00  
230194             JustEat  2023-12-06 21:32:00  

[230195 rows x 6 columns]

In [122]:
data['Third Party Account No'].value_counts() 

Third Party Account No
762896656.0    24
197865285.0    24
407574996.0    22
492058070.0    22
248658670.0    22
               ..
796767575.0     1
477605170.0     1
314289834.0     1
532478725.0     1
948933287.0     1
Name: count, Length: 673, dtype: int64

In [56]:
loan = data[data['Amount'] < -100]
loan = loan.sort_values(by=['Account No', 'DateTime'])

loan['DateTime'] = pd.to_datetime(loan['DateTime'])
loan['Date'] = loan['DateTime'].dt.date
loan['Time'] = loan['DateTime'].dt.time
loan

Account No      Balance  Amount  Third Party Account No  \
911     101531259.0   124.000000  -859.0                     NaN   
5127    101531259.0  -247.990000  -184.0                     NaN   
5194    101531259.0  -446.490000  -198.5                     NaN   
5566    101531259.0  -550.490000  -104.0                     NaN   
5621    101531259.0  -778.490000  -128.5                     NaN   
...             ...          ...     ...                     ...   
222249  999752672.0  2037.841094  -652.0                     NaN   
228192  999752672.0  1740.161094  -203.0                     NaN   
228262  999752672.0  1606.161094  -134.0                     NaN   
228329  999752672.0  1268.161094  -133.5                     NaN   
228352  999752672.0  1035.661094  -148.5                     NaN   

       Third Party Name            DateTime        Date      Time  
911                 LBG 2023-01-01 00:00:00  2023-01-01  00:00:00  
5127            Matalan 2023-01-01 12:00:00  2023-01-01  12:00:00  
5194            Topshop 2023-01-01 12:28:00  2023-01-01  12:28:00  
5566            Matalan 2023-01-01 15:54:00  2023-01-01  15:54:00  
5621            Matalan 2023-01-01 16:52:00  2023-01-01  16:52:00  
...                 ...                 ...         ...       ...  
222249          Halifax 2023-12-01 00:00:00  2023-12-01  00:00:00  
228192         Fat Face 2023-12-03 12:24:00  2023-12-03  12:24:00  
228262   Loosely Fitted 2023-12-03 13:32:00  2023-12-03  13:32:00  
228329          Matalan 2023-12-03 15:37:00  2023-12-03  15:37:00  
228352          Topshop 2023-12-03 16:34:00  2023-12-03  16:34:00  

[39390 rows x 8 columns]

In [57]:
loan['Time'].value_counts()

Time
00:00:00    6121
23:59:00    2286
17:37:00      87
17:53:00      83
17:58:00      82
            ... 
06:57:00       1
01:26:00       1
03:06:00       1
03:20:00       1
05:03:00       1
Name: count, Length: 1076, dtype: int64

In [58]:
condition1 = loan['Time'].isin([pd.to_datetime('00:00').time(), pd.to_datetime('23:59').time()])
condition2 = ~loan['Time'].isin([pd.to_datetime('00:00').time(), pd.to_datetime('23:59').time()])
loan1 = loan[condition1]
loan1

Account No      Balance  Amount  Third Party Account No  \
911     101531259.0   124.000000  -859.0                     NaN   
22239   101531259.0  -466.532050  -859.0                     NaN   
40633   101531259.0  -203.403250  -859.0                     NaN   
58193   101531259.0   193.085050  -859.0                     NaN   
76369   101531259.0   324.975700  -859.0                     NaN   
...             ...          ...     ...                     ...   
177813  999752672.0  1437.420944  -652.0                     NaN   
198501  999752672.0  2657.786894  -680.0             211493028.0   
200260  999752672.0  2005.786894  -652.0                     NaN   
220369  999752672.0  2839.263894  -680.0             211493028.0   
222249  999752672.0  2037.841094  -652.0                     NaN   

       Third Party Name            DateTime        Date      Time  
911                 LBG 2023-01-01 00:00:00  2023-01-01  00:00:00  
22239               LBG 2023-02-01 00:00:00  2023-02-01  00:00:00  
40633               LBG 2023-03-01 00:00:00  2023-03-01  00:00:00  
58193               LBG 2023-04-01 00:00:00  2023-04-01  00:00:00  
76369               LBG 2023-05-01 00:00:00  2023-05-01  00:00:00  
...                 ...                 ...         ...       ...  
177813          Halifax 2023-10-01 00:00:00  2023-10-01  00:00:00  
198501              NaN 2023-10-30 23:59:00  2023-10-30  23:59:00  
200260          Halifax 2023-11-01 00:00:00  2023-11-01  00:00:00  
220369              NaN 2023-11-29 23:59:00  2023-11-29  23:59:00  
222249          Halifax 2023-12-01 00:00:00  2023-12-01  00:00:00  

[8407 rows x 8 columns]

In [59]:
loan2 = loan[condition2]
loan2

Account No      Balance  Amount  Third Party Account No  \
5127    101531259.0  -247.990000 -184.00                     NaN   
5194    101531259.0  -446.490000 -198.50                     NaN   
5566    101531259.0  -550.490000 -104.00                     NaN   
5621    101531259.0  -778.490000 -128.50                     NaN   
11814   101531259.0 -1071.915850 -200.00                     NaN   
...             ...          ...     ...                     ...   
221602  999752672.0  2689.841094 -142.52                     NaN   
228192  999752672.0  1740.161094 -203.00                     NaN   
228262  999752672.0  1606.161094 -134.00                     NaN   
228329  999752672.0  1268.161094 -133.50                     NaN   
228352  999752672.0  1035.661094 -148.50                     NaN   

       Third Party Name            DateTime        Date      Time  
5127            Matalan 2023-01-01 12:00:00  2023-01-01  12:00:00  
5194            Topshop 2023-01-01 12:28:00  2023-01-01  12:28:00  
5566            Matalan 2023-01-01 15:54:00  2023-01-01  15:54:00  
5621            Matalan 2023-01-01 16:52:00  2023-01-01  16:52:00  
11814              Head 2023-01-13 03:00:00  2023-01-13  03:00:00  
...                 ...                 ...         ...       ...  
221602            Tesco 2023-11-30 19:50:00  2023-11-30  19:50:00  
228192         Fat Face 2023-12-03 12:24:00  2023-12-03  12:24:00  
228262   Loosely Fitted 2023-12-03 13:32:00  2023-12-03  13:32:00  
228329          Matalan 2023-12-03 15:37:00  2023-12-03  15:37:00  
228352          Topshop 2023-12-03 16:34:00  2023-12-03  16:34:00  

[30983 rows x 8 columns]

# Loan1 - All transaction in fix time happen 11~12 times

In [81]:
loan_value = loan1.groupby(['Account No', 'Amount'])['Amount'].count().reset_index(name='Count')
loan_value.columns = ['Account No', 'Amount', 'Count']
loan_value

Account No  Amount  Count
0    101531259.0  -859.0     12
1    104832000.0  -861.0     12
2    104832000.0  -670.0     12
3    108931112.0  -940.0     11
4    108931112.0  -936.0     12
..           ...     ...    ...
705  996042490.0 -1500.0     11
706  996042490.0  -501.0     12
707  998390769.0 -1013.0     12
708  999752672.0  -680.0     11
709  999752672.0  -652.0     12

[710 rows x 3 columns]

In [66]:
loan_value[loan_value['Count'] > 10]

Account No  Amount  Count
0    101531259.0  -859.0     12
1    104832000.0  -861.0     12
2    104832000.0  -670.0     12
3    108931112.0  -940.0     11
4    108931112.0  -936.0     12
..           ...     ...    ...
705  996042490.0 -1500.0     11
706  996042490.0  -501.0     12
707  998390769.0 -1013.0     12
708  999752672.0  -680.0     11
709  999752672.0  -652.0     12

[709 rows x 3 columns]

In [93]:
loan_value[loan_value['Count'] > 12]

Account No  Amount  Count
216  386729293.0  -650.0     23

In [94]:
loan1[loan1['Account No'] == 386729293]

Account No       Balance  Amount  Third Party Account No  \
544     386729293.0    313.000000  -650.0                     NaN   
20309   386729293.0   5116.631300  -650.0             391559884.0   
22420   386729293.0   4408.651300  -650.0                     NaN   
38660   386729293.0   8954.913687  -650.0             391559884.0   
40373   386729293.0   8304.913687  -650.0                     NaN   
56558   386729293.0  12705.765726  -650.0             391559884.0   
58059   386729293.0  12055.765726  -650.0                     NaN   
74694   386729293.0  16403.739897  -650.0             391559884.0   
76486   386729293.0  15696.837221  -650.0                     NaN   
93172   386729293.0  20002.192181  -650.0             391559884.0   
95160   386729293.0  19352.192181  -650.0                     NaN   
112728  386729293.0  22227.355357  -650.0             391559884.0   
114197  386729293.0  21577.355357  -650.0                     NaN   
133435  386729293.0  25894.758059  -650.0             391559884.0   
134960  386729293.0  25244.758059  -650.0                     NaN   
154348  386729293.0  27559.000668  -650.0             391559884.0   
156097  386729293.0  26876.060371  -650.0                     NaN   
176118  386729293.0  31252.571562  -650.0             391559884.0   
177843  386729293.0  30527.601562  -650.0                     NaN   
198489  386729293.0  35082.697899  -650.0             391559884.0   
200335  386729293.0  34366.566237  -650.0                     NaN   
220487  386729293.0  36445.164075  -650.0             391559884.0   
222524  386729293.0  35795.164075  -650.0                     NaN   

       Third Party Name            DateTime        Date      Time  
544                 LBG 2023-01-01 00:00:00  2023-01-01  00:00:00  
20309               NaN 2023-01-30 23:59:00  2023-01-30  23:59:00  
22420               LBG 2023-02-01 00:00:00  2023-02-01  00:00:00  
38660               NaN 2023-02-27 23:59:00  2023-02-27  23:59:00  
40373               LBG 2023-03-01 00:00:00  2023-03-01  00:00:00  
56558               NaN 2023-03-30 23:59:00  2023-03-30  23:59:00  
58059               LBG 2023-04-01 00:00:00  2023-04-01  00:00:00  
74694               NaN 2023-04-29 23:59:00  2023-04-29  23:59:00  
76486               LBG 2023-05-01 00:00:00  2023-05-01  00:00:00  
93172               NaN 2023-05-30 23:59:00  2023-05-30  23:59:00  
95160               LBG 2023-06-01 00:00:00  2023-06-01  00:00:00  
112728              NaN 2023-06-29 23:59:00  2023-06-29  23:59:00  
114197              LBG 2023-07-01 00:00:00  2023-07-01  00:00:00  
133435              NaN 2023-07-30 23:59:00  2023-07-30  23:59:00  
134960              LBG 2023-08-01 00:00:00  2023-08-01  00:00:00  
154348              NaN 2023-08-30 23:59:00  2023-08-30  23:59:00  
156097              LBG 2023-09-01 00:00:00  2023-09-01  00:00:00  
176118              NaN 2023-09-29 23:59:00  2023-09-29  23:59:00  
177843              LBG 2023-10-01 00:00:00  2023-10-01  00:00:00  
198489              NaN 2023-10-30 23:59:00  2023-10-30  23:59:00  
200335              LBG 2023-11-01 00:00:00  2023-11-01  00:00:00  
220487              NaN 2023-11-29 23:59:00  2023-11-29  23:59:00  
222524              LBG 2023-12-01 00:00:00  2023-12-01  00:00:00

#### In fix time, only 1 are less than 11 times transacton

In [67]:
loan_value[loan_value['Count'] < 2]

Account No  Amount  Count
545  779946617.0  -220.0      1

#### 779946617 have 2 record, -1300 and 12 times, -220 and 1 time

In [71]:
loan_value['Account No'].nunique()

517

In [72]:
duplicates = loan_value[loan_value.duplicated('Account No', keep=False)].sort_values('Account No')
duplicates

Account No  Amount  Count
1    104832000.0  -861.0     12
2    104832000.0  -670.0     12
3    108931112.0  -940.0     11
4    108931112.0  -936.0     12
5    111137699.0  -670.0     12
..           ...     ...    ...
703  994947671.0  -590.0     12
705  996042490.0 -1500.0     11
706  996042490.0  -501.0     12
708  999752672.0  -680.0     11
709  999752672.0  -652.0     12

[386 rows x 3 columns]

In [73]:
loan1['Date'].value_counts()

Date
2023-01-01    600
2023-02-01    503
2023-10-01    502
2023-03-01    502
2023-04-01    502
2023-06-01    502
2023-11-01    502
2023-08-01    502
2023-09-01    502
2023-12-01    501
2023-07-01    501
2023-05-01    501
2023-01-30    109
2023-11-29    109
2023-10-30    109
2023-09-29    109
2023-08-30    109
2023-07-30    109
2023-06-29    109
2023-05-30    109
2023-04-29    109
2023-03-30    109
2023-02-27    109
2023-06-30     99
2023-10-31     99
2023-09-30     99
2023-08-31     99
2023-07-31     99
2023-05-31     99
2023-04-30     99
2023-03-31     99
2023-02-28     99
2023-11-30     98
2023-01-31     98
2023-03-11      1
Name: count, dtype: int64

#### There is 976 Accounts in origin dataset, and 517 Accounts have fixed monthly loan
#### In 517 Accounts, 325 Accounts only have one loan, 192 Accounts have 2 loan

# Bank - 502 accounts

In [77]:
loan1['Third Party Name'].value_counts()

Third Party Name
Halifax              3056
LBG                  2964
Frankie & Bennies       1
Name: count, dtype: int64

In [112]:
loan_company = loan1.dropna(subset=['Third Party Name']).copy()
loan_company.drop(['Third Party Account No', 'Balance', 'DateTime', 'Date', 'Time'], axis=1, inplace=True)
loan_company = loan_company[loan_company['Third Party Name'] != 'Frankie & Bennies']
loan_company.drop_duplicates(inplace=True)
loan_company

Account No  Amount Third Party Name
911  101531259.0  -859.0              LBG
591  104832000.0  -861.0              LBG
567  108931112.0  -936.0              LBG
598  111137699.0  -660.0              LBG
978  117273481.0  -897.0              LBG
..           ...     ...              ...
973  994947671.0  -590.0              LBG
901  995615876.0  -637.0              LBG
706  996042490.0  -501.0          Halifax
854  998390769.0 -1013.0          Halifax
916  999752672.0  -652.0          Halifax

[502 rows x 3 columns]

In [111]:
loan_company.drop('Amount', axis=1, inplace=True)
loan_company.drop('Third Party Name', axis=1, inplace=True)
loan_company.drop_duplicates(inplace=True)
loan_company

Account No
911  101531259.0
591  104832000.0
567  108931112.0
598  111137699.0
978  117273481.0
..           ...
973  994947671.0
901  995615876.0
706  996042490.0
854  998390769.0
916  999752672.0

[502 rows x 1 columns]

#### 502 Accounts have one loan from Halifax or LBG

# Personal transfer - One people have two accounts? or rent? family?

In [78]:
loan1['Third Party Account No'].value_counts() 

Third Party Account No
587923845.0    12
158766229.0    12
857997794.0    12
572384506.0    12
492850697.0    12
               ..
669503912.0    11
985188462.0    11
761272540.0    11
429235692.0    11
211493028.0    11
Name: count, Length: 208, dtype: int64

In [113]:
loan_person = loan1.dropna(subset=['Third Party Account No']).copy()
loan_person.drop(['Third Party Name', 'Balance', 'DateTime', 'Date', 'Time'], axis=1, inplace=True)
loan_person.drop_duplicates(inplace=True)
loan_person

Account No  Amount  Third Party Account No
94     104832000.0  -670.0             587923845.0
20680  108931112.0  -940.0             646840694.0
102    111137699.0  -670.0             165098840.0
39     119993184.0  -430.0             384497904.0
20300  121858569.0  -940.0             248658670.0
...            ...     ...                     ...
20480  990894155.0  -540.0             262767079.0
20682  992204045.0 -1150.0             815993333.0
20522  994947671.0  -600.0             845648250.0
20222  996042490.0 -1500.0             792845450.0
20433  999752672.0  -680.0             211493028.0

[208 rows x 3 columns]

In [119]:
salary_one = pd.read_csv('C:/Users/Demi/0-Mini Project/salary_onetypes.csv')
salary_one.dropna(subset=['Third Party Account No'], inplace=True)
salary_one.drop(['Unnamed: 0', 'Third Party Name'], axis=1, inplace=True)

salary_one = salary_one.rename(columns={'Account No': 'Money in'})
salary_one = salary_one.rename(columns={'Amount': 'Money'})
salary_one = salary_one.rename(columns={'Third Party Account No': 'Money out'})
salary_one

Money in   Money    Money out
2    105375973.0   900.0  709632525.0
5    108812033.0   830.0  386277162.0
12   117308635.0  1200.0  255814263.0
13   119943495.0   400.0  352899149.0
16   121113546.0   410.0  490790931.0
..           ...     ...          ...
773  985188462.0   830.0  593378953.0
778  987656636.0  2670.0  450896297.0
782  993238248.0   540.0  448279144.0
783  993663752.0   670.0  672994751.0
789  998405607.0   450.0  388862238.0

[194 rows x 3 columns]

In [120]:
joined_df = pd.merge(loan_person, salary_one, left_on=['Account No', 'Third Party Account No'], right_on=['Money out', 'Money in'])
joined_df

Account No  Amount  Third Party Account No     Money in   Money  \
0    111137699.0  -670.0             165098840.0  165098840.0   670.0   
1    119993184.0  -430.0             384497904.0  384497904.0   430.0   
2    121858569.0  -940.0             248658670.0  248658670.0   940.0   
3    124125096.0 -2110.0             514362267.0  514362267.0  2110.0   
4    130154096.0  -860.0             799861885.0  799861885.0   860.0   
..           ...     ...                     ...          ...     ...   
173  990894155.0  -540.0             262767079.0  262767079.0   540.0   
174  992204045.0 -1150.0             815993333.0  815993333.0  1150.0   
175  994947671.0  -600.0             845648250.0  845648250.0   600.0   
176  996042490.0 -1500.0             792845450.0  792845450.0  1500.0   
177  999752672.0  -680.0             211493028.0  211493028.0   680.0   

       Money out  
0    111137699.0  
1    119993184.0  
2    121858569.0  
3    124125096.0  
4    130154096.0  
..           ...  
173  990894155.0  
174  992204045.0  
175  994947671.0  
176  996042490.0  
177  999752672.0  

[178 rows x 6 columns]

#### There is 178 transaction transfer money between two accounts in every month  
#### For example, both 111137699 and 165098840 have a lot of shopping record, which I think should be consider as a person have two account or a family account?

In [121]:
joined_df.to_csv('joined_df.csv')